hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-common/src/main/java/org/apache/hadoop/yarn/server/federation/store/records/RouterStoreToken.java

```diff
@@ -22,6 +22,7 @@
 import org.apache.hadoop.yarn.security.client.YARNDelegationTokenIdentifier;
 import org.apache.hadoop.yarn.util.Records;
 
+import java.io.DataInput;
```

Comment:
Why do we need this? We didn't add anything using DataInput to this class.

Reply:
We used the `readFields` method, which requires `DataInput` parameters.

RouterStoreToken#readFields
```
 public abstract void readFields(DataInput in) throws IOException;
```

When we store `RouterStoreToken`, we convert token into a byte array, and then store it in `ZK`

```
private void storeOrUpdateRouterRMDT(RouterRMTokenRequest request,  boolean isUpdate)
      throws Exception {
    RouterStoreToken routerStoreToken  = request.getRouterStoreToken();
    String nodeCreatePath = getStoreTokenZNodePathByTokenRequest(request);
    LOG.debug("nodeCreatePath = {}, isUpdate = {}", nodeCreatePath, isUpdate);
    put(nodeCreatePath, routerStoreToken.toByteArray(), isUpdate);
 }
```

When we need to read data, we convert the bytes array stored in `ZK` into `DataInputStream`, and then let `RouterStoreToken` parse the `required fields` from it.

```
private RouterStoreToken getStoreTokenFromZK(String nodePath, boolean quiet)
      throws IOException {
    try {
      byte[] data = get(nodePath);
      if ((data == null) || (data.length == 0)) {
        return null;
      }
      ByteArrayInputStream bin = new ByteArrayInputStream(data);
      DataInputStream din = new DataInputStream(bin);
      RouterStoreToken storeToken = Records.newRecord(RouterStoreToken.class);
      storeToken.readFields(din);
      return storeToken;
    } catch (Exception ex) {
      if (!quiet) {
        LOG.error("No node in path [" + nodePath + "]");
      }
      throw new IOException(ex);
    }
  }
```

Reply:
Shouldn't that be covered in the impl and not in the super class?

Reply:
The method we completed in the impl defines the abstract method declaration in the abstract class.

`RouterStoreToken` is an abstract class, its impl class is `RouterStoreTokenPBImpl`, we have implemented these 2 methods.

RouterStoreToken#toByteArray and readFields

```
 public abstract byte[] toByteArray() throws IOException;

 public abstract void readFields(DataInput in) throws IOException;
```

RouterStoreTokenPBImpl#toByteArray and readFields

```
public byte[] toByteArray() throws IOException {
    return builder.build().toByteArray();
}

public void readFields(DataInput in) throws IOException {
    builder.mergeFrom((DataInputStream) in);
}
```